File: zy73133/coordinate_parser.py
from __future__ import annotations

import re
from typing import Tuple, Optional, List

from models import NormalizedCoordinates, CoordinateFormat


LAT_PATTERNS = [
    (re.compile(r'^(-?\d+\.?\d*)\s*°?\s*([NSns]?)$'), CoordinateFormat.DECIMAL),
    (re.compile(r'^(-?\d+)[°\s]\s*(\d+\.?\d*)\s*[\'′]?\s*([NSns]?)$'), CoordinateFormat.DDM),
    (re.compile(r'^(-?\d+)[°\s]\s*(\d+)[\'′\s]\s*(\d+\.?\d*)[\"″]?\s*([NSns]?)$'), CoordinateFormat.DMS),
]

LNG_PATTERNS = [
    (re.compile(r'^(-?\d+\.?\d*)\s*°?\s*([EWew]?)$'), CoordinateFormat.DECIMAL),
    (re.compile(r'^(-?\d+)[°\s]\s*(\d+\.?\d*)\s*[\'′]?\s*([EWew]?)$'), CoordinateFormat.DDM),
    (re.compile(r'^(-?\d+)[°\s]\s*(\d+)[\'′\s]\s*(\d+\.?\d*)[\"″]?\s*([EWew]?)$'), CoordinateFormat.DMS),
]


def _parse_single_value(raw: str, is_latitude: bool) -> Tuple[Optional[float], CoordinateFormat, List[str]]:
    notes: List[str] = []
    if raw is None:
        return None, CoordinateFormat.UNKNOWN, ["原始值为空"]

    cleaned = raw.strip().replace(' ', ' ')
    if not cleaned:
        return None, CoordinateFormat.UNKNOWN, ["原始值为空字符串"]

    patterns = LAT_PATTERNS if is_latitude else LNG_PATTERNS

    for pattern, fmt in patterns:
        m = pattern.match(cleaned)
        if not m:
            continue

        groups = m.groups()

        if fmt == CoordinateFormat.DECIMAL:
            value = float(groups[0])
            direction = groups[1].upper() if groups[1] else ''
            if direction in ('S', 'W') and value > 0:
                value = -value
                notes.append(f"检测到{direction}方向，数值取反")
            elif direction in ('N', 'E') and value < 0:
                value = abs(value)
                notes.append(f"检测到{direction}方向，负数取正")
            return value, fmt, notes

        elif fmt == CoordinateFormat.DDM:
            degrees = float(groups[0])
            minutes = float(groups[1])
            direction = groups[2].upper() if groups[2] else ''
            if minutes >= 60:
                notes.append(f"分({minutes})超过60，按60进制转换")
            value = abs(degrees) + minutes / 60.0
            if degrees < 0 or direction in ('S', 'W'):
                value = -value
            if direction:
                notes.append(f"按{direction}方向判定符号")
            return value, fmt, notes

        elif fmt == CoordinateFormat.DMS:
            degrees = float(groups[0])
            minutes = float(groups[1])
            seconds = float(groups[2])
            direction = groups[3].upper() if groups[3] else ''
            if minutes >= 60:
                notes.append(f"分({minutes})超过60，按60进制转换")
            if seconds >= 60:
                notes.append(f"秒({seconds})超过60，按60进制转换")
            value = abs(degrees) + minutes / 60.0 + seconds / 3600.0
            if degrees < 0 or direction in ('S', 'W'):
                value = -value
            if direction:
                notes.append(f"按{direction}方向判定符号")
            return value, fmt, notes

    raw_numeric = re.sub(r'[^0-9.\-]', '', cleaned)
    if raw_numeric and raw_numeric not in ('-', '.', '-.'):
        try:
            value = float(raw_numeric)
            notes.append(f"无法识别格式，按纯数值提取: {raw_numeric}")
            return value, CoordinateFormat.UNKNOWN, notes
        except ValueError:
            pass

    return None, CoordinateFormat.UNKNOWN, [f"无法解析坐标值: {cleaned}"]


def _validate_range(lat: float, lng: float) -> List[str]:
    issues: List[str] = []
    if not (-90 <= lat <= 90):
        issues.append(f"纬度 {lat} 超出有效范围 [-90, 90]")
    if not (-180 <= lng <= 180):
        issues.append(f"经度 {lng} 超出有效范围 [-180, 180]")
    return issues


def parse_coordinates(raw_latitude: str, raw_longitude: str) -> NormalizedCoordinates:
    lat_val, lat_fmt, lat_notes = _parse_single_value(raw_latitude, is_latitude=True)
    lng_val, lng_fmt, lng_notes = _parse_single_value(raw_longitude, is_latitude=False)

    all_notes = lat_notes + lng_notes

    if lat_val is not None and lng_val is not None:
        range_issues = _validate_range(lat_val, lng_val)
        all_notes.extend(range_issues)

    fmt = lat_fmt if lat_fmt == lng_fmt else (
        CoordinateFormat.DMS if lat_fmt == CoordinateFormat.DMS or lng_fmt == CoordinateFormat.DMS
        else CoordinateFormat.DDM if lat_fmt == CoordinateFormat.DDM or lng_fmt == CoordinateFormat.DDM
        else CoordinateFormat.DECIMAL if lat_fmt == CoordinateFormat.DECIMAL or lng_fmt == CoordinateFormat.DECIMAL
        else CoordinateFormat.UNKNOWN
    )

    if lat_fmt != lng_fmt:
        all_notes.append(
            f"经纬格式不一致: 纬度={lat_fmt.value}, 经度={lng_fmt.value}，统一按 {fmt.value} 记录"
        )

    return NormalizedCoordinates(
        latitude=lat_val if lat_val is not None else 0.0,
        longitude=lng_val if lng_val is not None else 0.0,
        original_format=fmt,
        raw_latitude=raw_latitude,
        raw_longitude=raw_longitude,
        parse_notes=all_notes
    )

File: zy73133/src/engine/coordinateParser.ts
import type { CoordinateFormat, NormalizedCoordinates } from './types';

interface LatPattern {
  re: RegExp;
  fmt: CoordinateFormat;
}

const LAT_PATTERNS: LatPattern[] = [
  { re: /^(-?\d+\.?\d*)\s*°?\s*([NSns]?)$/, fmt: 'decimal' },
  { re: /^(-?\d+)[°\s]\s*(\d+\.?\d*)\s*['′]?\s*([NSns]?)$/, fmt: 'ddm' },
  { re: /^(-?\d+)[°\s]\s*(\d+)['′\s]\s*(\d+\.?\d*)["″]?\s*([NSns]?)$/, fmt: 'dms' },
];

const LNG_PATTERNS: LatPattern[] = [
  { re: /^(-?\d+\.?\d*)\s*°?\s*([EWew]?)$/, fmt: 'decimal' },
  { re: /^(-?\d+)[°\s]\s*(\d+\.?\d*)\s*['′]?\s*([EWew]?)$/, fmt: 'ddm' },
  { re: /^(-?\d+)[°\s]\s*(\d+)['′\s]\s*(\d+\.?\d*)["″]?\s*([EWew]?)$/, fmt: 'dms' },
];

interface ParsedValue {
  value: number | null;
  fmt: CoordinateFormat;
  notes: string[];
}

function parseSingleValue(raw: string, isLatitude: boolean): ParsedValue {
  const notes: string[] = [];
  if (raw == null) {
    return { value: null, fmt: 'unknown', notes: ['原始值为空'] };
  }

  const cleaned = String(raw).trim();
  if (!cleaned) {
    return { value: null, fmt: 'unknown', notes: ['原始值为空字符串'] };
  }

  const patterns = isLatitude ? LAT_PATTERNS : LNG_PATTERNS;

  for (const { re, fmt } of patterns) {
    const m = cleaned.match(re);
    if (!m) continue;

    if (fmt === 'decimal') {
      const value = parseFloat(m[1]);
      if (Number.isNaN(value)) continue;
      const direction = (m[2] || '').toUpperCase();
      let v = value;
      if ((direction === 'S' || direction === 'W') && v > 0) {
        v = -v;
        notes.push(`检测到${direction}方向，数值取反`);
      } else if ((direction === 'N' || direction === 'E') && v < 0) {
        v = Math.abs(v);
        notes.push(`检测到${direction}方向，负数取正`);
      }
      return { value: v, fmt, notes };
    }

    if (fmt === 'ddm') {
      const degrees = parseFloat(m[1]);
      const minutes = parseFloat(m[2]);
      if (Number.isNaN(degrees) || Number.isNaN(minutes)) continue;
      const direction = (m[3] || '').toUpperCase();
      if (minutes >= 60) notes.push(`分(${minutes})超过60，按60进制转换`);
      let v = Math.abs(degrees) + minutes / 60;
      if (degrees < 0 || direction === 'S' || direction === 'W') v = -v;
      if (direction) notes.push(`按${direction}方向判定符号`);
      return { value: v, fmt, notes };
    }

    if (fmt === 'dms') {
      const degrees = parseFloat(m[1]);
      const minutes = parseFloat(m[2]);
      const seconds = parseFloat(m[3]);
      if (Number.isNaN(degrees) || Number.isNaN(minutes) || Number.isNaN(seconds)) continue;
      const direction = (m[4] || '').toUpperCase();
      if (minutes >= 60) notes.push(`分(${minutes})超过60，按60进制转换`);
      if (seconds >= 60) notes.push(`秒(${seconds})超过60，按60进制转换`);
      let v = Math.abs(degrees) + minutes / 60 + seconds / 3600;
      if (degrees < 0 || direction === 'S' || direction === 'W') v = -v;
      if (direction) notes.push(`按${direction}方向判定符号`);
      return { value: v, fmt, notes };
    }
  }

  // Fallback: try to extract a pure numeric value, but mark format unknown.
  const rawNumeric = cleaned.replace(/[^0-9.\-]/g, '');
  if (rawNumeric && !['-', '.', '-.'].includes(rawNumeric)) {
    const value = parseFloat(rawNumeric);
    if (!Number.isNaN(value)) {
      notes.push(`无法识别格式，按纯数值提取: ${rawNumeric}`);
      return { value, fmt: 'unknown', notes };
    }
  }

  return { value: null, fmt: 'unknown', notes: [`无法解析坐标值: ${cleaned}`] };
}

function validateRange(lat: number, lng: number): string[] {
  const issues: string[] = [];
  if (!(-90 <= lat && lat <= 90)) issues.push(`纬度 ${lat} 超出有效范围 [-90, 90]`);
  if (!(-180 <= lng && lng <= 180)) issues.push(`经度 ${lng} 超出有效范围 [-180, 180]`);
  return issues;
}

function pickOverallFormat(latFmt: CoordinateFormat, lngFmt: CoordinateFormat): CoordinateFormat {
  if (latFmt === lngFmt) return latFmt;
  if (latFmt === 'dms' || lngFmt === 'dms') return 'dms';
  if (latFmt === 'ddm' || lngFmt === 'ddm') return 'ddm';
  if (latFmt === 'decimal' || lngFmt === 'decimal') return 'decimal';
  return 'unknown';
}

export function parseCoordinates(rawLatitude: string, rawLongitude: string): NormalizedCoordinates {
  const latParsed = parseSingleValue(rawLatitude, true);
  const lngParsed = parseSingleValue(rawLongitude, false);

  const allNotes = [...latParsed.notes, ...lngParsed.notes];

  if (latParsed.value != null && lngParsed.value != null) {
    allNotes.push(...validateRange(latParsed.value, lngParsed.value));
  } else {
    allNotes.push('坐标解析失败：经纬度未全部识别为有效数值，点位不落海面');
  }

  const fmt = pickOverallFormat(latParsed.fmt, lngParsed.fmt);
  if (latParsed.fmt !== lngParsed.fmt) {
    allNotes.push(`经纬格式不一致: 纬度=${latParsed.fmt}, 经度=${lngParsed.fmt}，统一按 ${fmt} 记录`);
  }

  return {
    latitude: latParsed.value,
    longitude: lngParsed.value,
    originalFormat: fmt,
    rawLatitude: rawLatitude ?? '',
    rawLongitude: rawLongitude ?? '',
    parseNotes: allNotes,
  };
}
